### PR TITLE
refac: OutgoingMessage.Application/Domain

### PR DESCRIPTION
**Pull request by lasrinnil**
## Description
~~All of the `DocumentWriter` implementations were placed into a `MarketDocuments` namespace. So I moved the writes into the `DocumentWriters` namespace.~~

Worked on the above but ended up refactoring most of the OutgoingMessage module. Will look into the `infrastructure` in another PR.

New folderstructure
image

### CODE REVIEW

File: source/OutgoingMessages.Application/Usecases/Peek.cs [hunk 1]
```diff
@@ -19,14 +19,20 @@
...
 using Energinet.DataHub.EDI.OutgoingMessages.Interfaces.Models;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Application.Usecases;
 
-public class MessagePeeker
+/// <summary>
+/// Peek is used by the actor to peek at the next message in their queue.
+/// </summary>
+public class Peek
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Application/Usecases/Peek.cs:35]: Maybe `PeekMessage` or `PeekOutgoingMessage`?

File: source/ArchitectureTests/RegistrationTests.cs [hunk 2]
```diff
@@ -66,15 +68,15 @@ public RegistrationTests()
         public static IEnumerable<object[]> GetDocumentWriterRequirements()
         {
             var constructorDependencies = ReflectionHelper.FindAllConstructorDependenciesForType();
-            return typeof(NotifyAggregatedMeasureDataXmlDocumentWriter).Assembly.GetTypes()
+            return typeof(NotifyAggregatedMeasureDataXmlXmlDocumentWriter).Assembly.GetTypes()
```
Comment (ebbeknudsen) [inline source/ArchitectureTests/RegistrationTests.cs:71]: Is the double `Xml` in name intentional?

File: source/OutgoingMessages.Domain/DocumentWriters/Formats/CIM/Xml/XmlDocumentWriter.cs [hunk 3]
```diff
@@ -19,18 +19,17 @@
...
 using Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.OutgoingMessages;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application.DocumentWriters.Xml;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM.Xml;
 
-public abstract class DocumentWriter : IDocumentWriter
+public abstract class XmlDocumentWriter : IDocumentWriter
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/DocumentWriters/Formats/CIM/Xml/XmlDocumentWriter.cs:26]:
```suggestion
public abstract class CimXmlDocumentWriter : IDocumentWriter
```

File: source/OutgoingMessages.Domain/DocumentWriters/Formats/CIM/Xml/XmlHeaderWriter.cs [hunk 4]
```diff
@@ -16,10 +16,9 @@
...
 using System.Xml;
 using Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.OutgoingMessages;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application.DocumentWriters.Xml;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM.Xml;
 
 internal static class XmlHeaderWriter
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/DocumentWriters/Formats/CIM/Xml/XmlHeaderWriter.cs:23]:
```suggestion
internal static class CimXmlHeaderWriter
```

File: source/OutgoingMessages.Domain/DocumentWriters/NotifyAggregatedMeasureData/NotifyAggregatedMeasureDataXmlXmlDocumentWriter.cs [hunk 5]
```diff
@@ -18,14 +18,15 @@
...
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM.Xml;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application.MarketDocuments.NotifyAggregatedMeasureData;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.NotifyAggregatedMeasureData;
 
-public class NotifyAggregatedMeasureDataXmlDocumentWriter : DocumentWriter
+public class NotifyAggregatedMeasureDataXmlXmlDocumentWriter : XmlDocumentWriter
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/DocumentWriters/NotifyAggregatedMeasureData/NotifyAggregatedMeasureDataXmlXmlDocumentWriter.cs:27]:
```suggestion
public class NotifyAggregatedMeasureDataXmlDocumentWriter : XmlDocumentWriter
```

File: source/OutgoingMessages.Domain/DocumentWriters/NotifyWholesaleServices/NotifyWholesaleServicesXmlXmlDocumentWriter.cs [hunk 6]
```diff
@@ -18,14 +18,15 @@
...
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM.Xml;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application.MarketDocuments.NotifyWholesaleServices;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.NotifyWholesaleServices;
 
-public class NotifyWholesaleServicesXmlDocumentWriter : DocumentWriter
+public class NotifyWholesaleServicesXmlXmlDocumentWriter : XmlDocumentWriter
 {
-    public NotifyWholesaleServicesXmlDocumentWriter(IMessageRecordParser parser)
+    public NotifyWholesaleServicesXmlXmlDocumentWriter(IMessageRecordParser parser)
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/DocumentWriters/NotifyWholesaleServices/NotifyWholesaleServicesXmlXmlDocumentWriter.cs:29]:
```suggestion
    public NotifyWholesaleServicesXmlDocumentWriter(IMessageRecordParser parser)
```

File: source/OutgoingMessages.Domain/DocumentWriters/RejectRequestAggregatedMeasureData/RejectRequestAggregatedMeasureDataXmlXmlDocumentWriter.cs [hunk 7]
```diff
@@ -17,14 +17,14 @@
...
-using Energinet.DataHub.EDI.OutgoingMessages.Application.DocumentWriters.Xml;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM.Xml;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application.MarketDocuments.RejectRequestAggregatedMeasureData;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.RejectRequestAggregatedMeasureData;
 
-public class RejectRequestAggregatedMeasureDataXmlDocumentWriter : DocumentWriter
+public class RejectRequestAggregatedMeasureDataXmlXmlDocumentWriter : XmlDocumentWriter
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/DocumentWriters/RejectRequestAggregatedMeasureData/RejectRequestAggregatedMeasureDataXmlXmlDocumentWriter.cs:25]:
```suggestion
public class RejectRequestAggregatedMeasureDataXmlDocumentWriter : XmlDocumentWriter
```

File: source/OutgoingMessages.Domain/DocumentWriters/RejectRequestWholesaleSettlement/RejectRequestWholesaleSettlementXmlXmlDocumentWriter.cs [hunk 8]
```diff
@@ -17,14 +17,14 @@
...
-using Energinet.DataHub.EDI.OutgoingMessages.Application.DocumentWriters.Xml;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM.Xml;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application.MarketDocuments.RejectRequestWholesaleSettlement;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.RejectRequestWholesaleSettlement;
 
-public class RejectRequestWholesaleSettlementXmlDocumentWriter : DocumentWriter
+public class RejectRequestWholesaleSettlementXmlXmlDocumentWriter : XmlDocumentWriter
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/DocumentWriters/RejectRequestWholesaleSettlement/RejectRequestWholesaleSettlementXmlXmlDocumentWriter.cs:25]:
```suggestion
public class RejectRequestWholesaleSettlementXmlDocumentWriter : XmlDocumentWriter
```

File: source/OutgoingMessages.Domain/Queueing/Bundels/Bundle.cs [hunk 9]
```diff
@@ -13,9 +13,10 @@
 // limitations under the License.
 
 using Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.OutgoingMessages;
 using NodaTime;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.Bundels;
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/Queueing/Bundels/Bundle.cs:19]:
```suggestion
namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.Bundles;
```

File: source/OutgoingMessages.Domain/Queueing/Bundels/BundleId.cs [hunk 10]
```diff
@@ -14,7 +14,7 @@
 
 using System;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.Bundels;
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/Queueing/Bundels/BundleId.cs:17]:
```suggestion
namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.Bundles;
```

**Review by ebbeknudsen — CHANGES_REQUESTED** [review]

File: source/OutgoingMessages.Domain/Queueing/OutgoingMessages/IOutgoingMessageRepository.cs [hunk 11]
```diff
@@ -13,9 +13,9 @@
 // limitations under the License.
 
 using System.Threading.Tasks;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.Bundels;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.OutgoingMessages
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/Queueing/OutgoingMessages/IOutgoingMessageRepository.cs:18]: Does the `Queueing` folder make sense? Maybe we should just move the `Bundels` (`Bundles`) and the `OutgoingMessages` into the root folder

File: source/ArchitectureTests/RegistrationTests.cs (hunk 2 above)
Comment (lasrinnil) [reply source/ArchitectureTests/RegistrationTests.cs:71]: This is Rsharper, good catch ill fit it. Only one XML is needed :)

File: source/OutgoingMessages.Application/Usecases/Delegation.cs [hunk 12]
```diff
@@ -17,18 +17,19 @@
...
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Queueing.OutgoingMessages;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Application.Usecases;
 
 /// <summary>
 /// Responsible for delegating messages to the correct receiver if a delegation relationship exists.
 /// </summary>
-public class MessageDelegator
+public class Delegation
```
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/OutgoingMessages.Application/Usecases/Delegation.cs:28]: Lets go with the <usecase>message approch 👍

File: source/OutgoingMessages.Domain/Queueing/OutgoingMessages/IOutgoingMessageRepository.cs (hunk 11 above)
Comment (lasrinnil) [reply source/OutgoingMessages.Domain/Queueing/OutgoingMessages/IOutgoingMessageRepository.cs]: how about this change:
image

File: source/OutgoingMessages.Domain/Models/Bundels/Bundle.cs [hunk 13]
```diff
@@ -13,9 +13,10 @@
 // limitations under the License.
 
 using Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.OutgoingMessages;
 using NodaTime;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.Bundels;
```
Comment (ebbeknudsen) [inline source/OutgoingMessages.Domain/Models/Bundels/Bundle.cs:19]: Seems like there is a typo in the folder name: `Bundels` -> `Bundles`

**Review by ebbeknudsen — APPROVED** [review]

File: source/OutgoingMessages.Domain/Queueing/OutgoingMessages/IOutgoingMessageRepository.cs (hunk 11 above)
Comment (ebbeknudsen) [reply source/OutgoingMessages.Domain/Queueing/OutgoingMessages/IOutgoingMessageRepository.cs]: Looks good to me (maybe except the `Bundels` -> `Bundles` typo 😄)

File: source/OutgoingMessages.Domain/Models/ReceiverMismatchException.cs [hunk 14]
```diff
@@ -14,7 +14,7 @@
 
 using System;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.Models;
```
Comment (MadsDue) [inline source/OutgoingMessages.Domain/Models/ReceiverMismatchException.cs:17]: Should we move this in to the Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.ActorMessagesQueues namespace, that is the only place it is use

File: source/OutgoingMessages.Domain/DocumentFactory.cs [hunk 15]
```diff
@@ -17,13 +17,12 @@
...
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters;
 using Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Exceptions;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.OutgoingMessages;
 using NodaTime;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain;
```
Comment (MadsDue) [inline source/OutgoingMessages.Domain/DocumentFactory.cs:25]: Consider moving this to Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters

File: source/OutgoingMessages.Domain/IMessageRecordParser.cs [hunk 16]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain
```
Comment (MadsDue) [inline source/OutgoingMessages.Domain/IMessageRecordParser.cs:15]: Consider moving this to Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters

File: source/OutgoingMessages.Domain/MarketDocumentWriterMemoryStream.cs [hunk 17]
```diff
@@ -14,6 +14,6 @@
 
 using System.IO;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain;
```
Comment (MadsDue) [inline source/OutgoingMessages.Domain/MarketDocumentWriterMemoryStream.cs:17]: Consider moving this to Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters

File: source/OutgoingMessages.Domain/MessageRecordParser.cs [hunk 18]
```diff
@@ -13,9 +13,8 @@
 // limitations under the License.
 
 using Energinet.DataHub.EDI.BuildingBlocks.Interfaces;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain
```
Comment (MadsDue) [inline source/OutgoingMessages.Domain/MessageRecordParser.cs:17]: Consider moving this to Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters

**Review by MadsDue — COMMENTED** [review]
I am still not sure about the interfaces for the repositories. 

Should the domain layer be able to trigger execute I/O operations?
 
Currently it looks like the repositories is only used in the application layer, which I also would except.

File: source/OutgoingMessages.Application/UseCases/DelegateMessage.cs [hunk 19]
```diff
@@ -17,18 +17,19 @@
...
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Models;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.OutgoingMessages;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Application.UseCases;
 
 /// <summary>
 /// Responsible for delegating messages to the correct receiver if a delegation relationship exists.
 /// </summary>
-public class MessageDelegator
```
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/OutgoingMessages.Application/UseCases/DelegateMessage.cs:27]: I could be! But I like that we have encapsulated it in its own "usecase" this way it's clear how we have implemented this particulier businesscase.

File: source/OutgoingMessages.Application/UseCases/DelegateMessage.cs [hunk 20]
```diff
@@ -17,18 +17,19 @@
 using System.Threading.Tasks;
 using Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
 using Energinet.DataHub.EDI.MasterData.Interfaces;
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.ActorMessagesQueues;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.OutgoingMessages;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Application.UseCases;
```
Comment (RasmusGraabaek) [inline source/OutgoingMessages.Application/UseCases/DelegateMessage.cs:23]:
```suggestion
namespace Energinet.DataHub.EDI.OutgoingMessages.Application.Features;
```

Comment (RasmusGraabaek) [reply source/OutgoingMessages.Application/UseCases/DelegateMessage.cs:23]: Maybe?

File: source/OutgoingMessages.Domain/DocumentWriters/NotifyAggregatedMeasureData/NotifyAggregatedMeasureDataJsonDocumentWriter.cs [hunk 21]
```diff
@@ -19,12 +19,12 @@
...
-using Energinet.DataHub.EDI.OutgoingMessages.Domain.OutgoingMessages.Queueing;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.Formats.CIM.Json;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.MarketDocuments;
+using Energinet.DataHub.EDI.OutgoingMessages.Domain.Models.OutgoingMessages;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application.MarketDocuments.NotifyAggregatedMeasureData;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.NotifyAggregatedMeasureData;
 
 public class NotifyAggregatedMeasureDataJsonDocumentWriter : IDocumentWriter
```
Comment (RasmusGraabaek) [inline source/OutgoingMessages.Domain/DocumentWriters/NotifyAggregatedMeasureData/NotifyAggregatedMeasureDataJsonDocumentWriter.cs:29]:
```suggestion
public class NotifyAggregatedMeasureDataCIMJsonDocumentWriter : IDocumentWriter
```
?

Comment (RasmusGraabaek) [reply source/OutgoingMessages.Domain/DocumentWriters/NotifyAggregatedMeasureData/NotifyAggregatedMeasureDataJsonDocumentWriter.cs:29]: This suggestion holds for all json stuff

Comment (lasrinnil) [reply source/OutgoingMessages.Domain/DocumentWriters/NotifyAggregatedMeasureData/NotifyAggregatedMeasureDataJsonDocumentWriter.cs]: done

File: source/OutgoingMessages.Application/UseCases/DelegateMessage.cs (hunk 20 above)
Comment (lasrinnil) [reply source/OutgoingMessages.Application/UseCases/DelegateMessage.cs:23]: No.

File: source/OutgoingMessages.Domain/DocumentWriters/NotifyWholesaleServices/WholesaleCalculationMarketActivityRecord.cs [hunk 22]
```diff
@@ -16,7 +16,7 @@
 using System.Collections.Generic;
 using Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
 
-namespace Energinet.DataHub.EDI.OutgoingMessages.Application.MarketDocuments.NotifyWholesaleServices;
+namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.DocumentWriters.NotifyWholesaleServices;
```
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/OutgoingMessages.Domain/DocumentWriters/NotifyWholesaleServices/WholesaleCalculationMarketActivityRecord.cs:19]: Its one of the serialized classes - so I would prefer not to :)

**Review by RasmusGraabaek — APPROVED** [review]

**Pull request merged**